### ISSUE TITLE
No configurator for root element <credentials>

### ISSUE THREAD

**Issue by jcrben** [jenkinsci/configuration-as-code-plugin#523]
I'm seeing this error and I can't figure it out. I have my minimal reproduction repo using the demo credentials config: https://gitlab.com/jcrben-not-mine/jenkins/casc-docker/tree/credentials-configurator

If you comment out https://gitlab.com/jcrben-not-mine/jenkins/casc-docker/blob/credentials-configurator/jenkins.yaml#L78-88 it should throw the error

Note that I do have the configuration-as-code-support plugin installed.

The master branch running 0.11 works.

Incidentally, I couldn't get the demo config for build-agents to work either (threw a NullPointerException - think it's the empty jnlp launcher). Would love to see a working configuration that someone on the team is building.

**Comment by casz** [comment]
This error? with no stacktrace?
![Come on, son](https://media.giphy.com/media/89No6UHzuUam4/giphy.gif)

**Comment by jcrben** [comment]
Most people don't post complete reproductions that you can build with a two commands - I don't think I've seen anyone else post such a thing in chat or here, including the team  😄 
```
./builder_jenkins.sh
docker-compose up -d
```

With that said, I can post a stack trace but it's not exactly going to be that illuminating - this is a not uncommon error as I mentioned above and the stack is going to look pretty much the same as the other instances of the error (https://github.com/jenkinsci/configuration-as-code-plugin/issues?q=is%3Aissue+No+configurator+for+root+element+is%3Aopen)

**Comment by casz** [comment]
To me a stacktrace is just as valuable as the issue and repro, the chance of me checking out the code is less 😅

**Comment by casz** [comment]
@jcrben dno if your plugin list is good enough: You definitely need https://plugins.jenkins.io/credentials and https://plugins.jenkins.io/plain-credentials

**Comment by jcrben** [comment]
Hmm, well now I'm having trouble reproducing. Even without that plain-credentials plugin explicitly listed - plain-credentials is a dependency of ssh-credentials and generally it is being pulled in...

I had alphabetized my plugins list which pulled configuration-as-code up to the top of the list, so thinking it was a chicken and egg problem, but now it's working even with configuration-as-code high in the list.

**Comment by jcrben** [comment]
That is fair, there's a couple dependencies! Thanks for looking at this.

It's not reproducing any longer anyway. My macbook was having issues at the time - docker was running out of space - and I ended up having to wipe docker for mac clean, so maybe that had something to do with it. Some sort of race condition causing a slowdown in some important spot. I guess the race condition is already logged via that chicken and egg issue, so I'll close this.

**Comment by jcrben** [comment]
So I figured this out. It was caused by:
`matrix-auth:2.2`
which needed to be upgraded
`matrix-auth:2.3`

**Comment by timja** [comment]
@casz
I hit this issue too

Removing matrix auth plugin has allowed me to work on this again.

Stack trace:
```
jenkins    | WARNING: Error while serving http://localhost:8088/configuration-as-code/reload
jenkins    | java.lang.reflect.InvocationTargetException
jenkins    | 	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:347)
jenkins    | 	at org.kohsuke.stapler.interceptor.RequirePOST$Processor.invoke(RequirePOST.java:77)
jenkins    | 	at org.kohsuke.stapler.PreInvokeInterceptedFunction.invoke(PreInvokeInterceptedFunction.java:26)
jenkins    | 	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:184)
jenkins    | 	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:117)
jenkins    | 	at org.kohsuke.stapler.MetaClass$1.doDispatch(MetaClass.java:129)
jenkins    | 	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
jenkins    | 	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:734)
jenkins    | 	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:864)
jenkins    | 	at org.kohsuke.stapler.MetaClass$10.dispatch(MetaClass.java:374)
jenkins    | 	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:734)
jenkins    | 	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:864)
jenkins    | 	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:668)
jenkins    | 	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
jenkins    | 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
jenkins    | 	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865)
jenkins    | 	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1655)
jenkins    | 	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:154)
jenkins    | 	at org.jenkinsci.plugins.ssegateway.Endpoint$SSEListenChannelFilter.doFilter(Endpoint.java:243)
jenkins    | 	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
jenkins    | 	at io.jenkins.blueocean.ResourceCacheControl.doFilter(ResourceCacheControl.java:134)
jenkins    | 	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
jenkins    | 	at io.jenkins.blueocean.auth.jwt.impl.JwtAuthenticationFilter.doFilter(JwtAuthenticationFilter.java:61)
jenkins    | 	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
jenkins    | 	at com.smartcodeltd.jenkinsci.plugin.assetbundler.filters.LessCSS.doFilter(LessCSS.java:47)
jenkins    | 	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
jenkins    | 	at hudson.plugins.greenballs.GreenBallFilter.doFilter(GreenBallFilter.java:59)
jenkins    | 	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
jenkins    | 	at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:239)
jenkins    | 	at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:215)
jenkins    | 	at net.bull.javamelody.PluginMonitoringFilter.doFilter(PluginMonitoringFilter.java:88)
jenkins    | 	at org.jvnet.hudson.plugins.monitoring.HudsonMonitoringFilter.doFilter(HudsonMonitoringFilter.java:114)
jenkins    | 	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
jenkins    | 	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:157)
jenkins    | 	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
jenkins    | 	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:99)
jenkins    | 	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
jenkins    | 	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
jenkins    | 	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:90)
jenkins    | 	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
jenkins    | 	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
jenkins    | 	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
jenkins    | 	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
jenkins    | 	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
jenkins    | 	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
jenkins    | 	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
jenkins    | 	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
jenkins    | 	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
jenkins    | 	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
jenkins    | 	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
jenkins    | 	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
jenkins    | 	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
jenkins    | 	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
jenkins    | 	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
jenkins    | 	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1317)
jenkins    | 	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
jenkins    | 	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
jenkins    | 	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
jenkins    | 	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
jenkins    | 	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1219)
jenkins    | 	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
jenkins    | 	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
jenkins    | 	at org.eclipse.jetty.server.Server.handle(Server.java:531)
jenkins    | 	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
jenkins    | 	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
jenkins    | 	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
jenkins    | 	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
jenkins    | 	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
jenkins    | 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
jenkins    | 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
jenkins    | 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
jenkins    | 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
jenkins    | 	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
jenkins    | 	at winstone.BoundedExecutorService.lambda$scheduleNext$0(BoundedExecutorService.java:80)
jenkins    | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
jenkins    | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
jenkins    | 	at java.lang.Thread.run(Thread.java:748)
jenkins    | Caused by: io.jenkins.plugins.casc.ConfiguratorException: No configurator for root element <credentials>
jenkins    | 	at io.jenkins.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:620)
jenkins    | 	at io.jenkins.plugins.casc.ConfigurationAsCode.checkWith(ConfigurationAsCode.java:641)
jenkins    | 	at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:628)
jenkins    | 	at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:540)
jenkins    | 	at io.jenkins.plugins.casc.ConfigurationAsCode.configure(ConfigurationAsCode.java:270)
jenkins    | 	at io.jenkins.plugins.casc.ConfigurationAsCode.doReload(ConfigurationAsCode.java:151)
jenkins    | 	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
jenkins    | 	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:343)
jenkins    | 	... 76 more
```

**Comment by casz** [comment]
@timja your missing configuration-as-code-support plugin 😅

**Comment by timja** [comment]
Don't think so
This is my plugins.txt file:

```
ace-editor:1.1
ansicolor:0.5.2
antisamy-markup-formatter:1.5
apache-httpcomponents-client-4-api:4.5.5-3.0
authentication-tokens:1.3
azure-commons:0.2.8
azure-credentials:1.6.0
azure-vm-agents:0.7.3
blueocean-autofavorite:1.2.2
blueocean-bitbucket-pipeline:1.8.4
blueocean-commons:1.8.4
blueocean-config:1.8.4
blueocean-core-js:1.8.4
blueocean-dashboard:1.8.4
blueocean-display-url:2.2.0
blueocean-events:1.8.4
blueocean-git-pipeline:1.8.4
blueocean-github-pipeline:1.8.4
blueocean-i18n:1.8.4
blueocean-jira:1.8.4
blueocean-jwt:1.8.4
blueocean-personalization:1.8.4
blueocean-pipeline-api-impl:1.8.4
blueocean-pipeline-editor:1.8.4
blueocean-pipeline-scm-api:1.8.4
blueocean-rest-impl:1.8.4
blueocean-rest:1.8.4
blueocean-web:1.8.4
blueocean:1.8.4
bouncycastle-api:2.17
branch-api:2.0.20
build-monitor-plugin:1.12+build.201809061734
cloud-stats:0.19
cloudbees-bitbucket-branch-source:2.2.12
cloudbees-folder:6.6
command-launcher:1.2
configuration-as-code:1.0
configuration-as-code-support:1.0
credentials-binding:1.16
credentials:2.1.18
custom-tools-plugin:0.5
display-url-api:2.2.0
docker-commons:1.13
docker-workflow:1.17
durable-task:1.26
extended-choice-parameter:0.76
external-monitor-job:1.7
favorite:2.3.2
gatling:1.2.5
git-client:2.7.3
git-server:1.7
git:3.9.1
github-api:1.92
github-branch-source:2.3.6
github:1.29.2
greenballs:1.15
handlebars:1.1.1
handy-uri-templates-2-api:2.1.6-1.0
htmlpublisher:1.16
http_request:1.8.22
jackson2-api:2.8.11.3
jacoco:3.0.3
javadoc:1.4
jdk-tool:1.1
jenkins-design-language:1.8.4
jira:3.0.2
job-dsl:1.70
jquery-detached:1.2.1
jquery:1.12.4-0
jsch:0.1.54.2
junit:1.26.1
lockable-resources:2.3
mailer:1.21
maven-plugin:3.1.2
mercurial:2.4
momentjs:1.1.1
monitoring:1.74.0
pam-auth:1.4
pipeline-build-step:2.7
pipeline-github-lib:1.0
pipeline-graph-analysis:1.7
pipeline-input-step:2.8
pipeline-milestone-step:1.3.1
pipeline-model-api:1.3.2
pipeline-model-declarative-agent:1.1.1
pipeline-model-definition:1.3.2
pipeline-model-extensions:1.3.2
pipeline-rest-api:2.10
pipeline-stage-step:2.3
pipeline-stage-tags-metadata:1.3.2
pipeline-stage-view:2.10
pipeline-utility-steps:2.1.0
plain-credentials:1.4
pubsub-light:1.12
run-condition:1.2
saml:1.0.9
sauce-ondemand:1.177
scm-api:2.2.8
script-security:1.46
slack:2.3
sonar:2.8.1
sse-gateway:1.16
ssh-credentials:1.14
structs:1.15
timestamper:1.8.10
token-macro:2.5
variant:1.1
windows-slaves:1.3.1
workflow-aggregator:2.5
workflow-api:2.29
workflow-basic-steps:2.11
workflow-cps-global-lib:2.11
workflow-cps:2.56
workflow-durable-task-step:2.22
workflow-job:2.25
workflow-multibranch:2.20
workflow-scm-step:2.7
workflow-step-api:2.16
workflow-support:2.20
```

**Comment by relief-melone** [comment]
Just in case someone landed here with the error message and is using a config as code version >= 1.18 check the releases

https://github.com/jenkinsci/configuration-as-code-plugin/releases

config as code support plugin is no longer needed. However you should include the credentials plugin if you get this error message

**Comment by casz** [comment]
Are the release notes not clear enough, I am open to suggestions? @relief-melone :cry:

**Comment by relief-melone** [comment]
I guess they should be. Maybe it was just in my case (i tend to use latest on all my plugins as I prefer solving a problem from time to time with a new version instead of piling up an update monster) that my running setup stopped working at once and it took me a while to figure out that a new build had actually pulled newer versions of casc.

So I went with the lazy googling the error message method and ended up here. After figuring out what went wrong I just thought it might be helpful to others.

I guess no suggestions though as the release notes state pretty clearly what was changed and it was pretty easy to fix my errors after I found it.
